### DISCUSSION TITLE
relation to `xdggs`?

### DISCUSSION THREAD

**Discussion by keewis** [UXARRAY/uxarray#1055] [General]
Over the past year, I've been working on [`xdggs`](https://github.com/xarray-contrib/xdggs), a package that allow working with discrete global grid systems (DGGS) by making use of the information in the indexing scheme.

As far as I understand it, `uxarray` works in a similar space (for example see #1028), but instead of using the cell / node / zone ids it makes use of the mesh / cell boundaries.

In opening this discussion I'm particularly interested in figuring out how both libraries relate to each other, how they could interact (e.g. transformation between the two in-memory representations), and in general how we can collaborate better.

cc @allixender

**Comment by keewis** [comment]
To be more concrete for conversion of the in-memory representations, do you think we could use `geoarrow` (or `shapely`) polygon arrays to convert the `xr.Dataset` object to `uxarray`? If I read your documentation correctly, we'd have to somehow construct a `uxarray.Grid` from the polygon arrays and would then be able to use `UxDataset.from_xarray`.

For the other way around, if the cell ids / grid parameters are already available I think it might be fine to convert to `xarray` and use the usual `xdggs.decode`. However, if that's not the case we'd have to figure out the grid parameters (mostly resolution level, but depending on the grid there might be more), and then we'd be able to use the cell centers to reconstruct the cell ids.

Once we've resolved those question I guess we'd also have to figure out where the conversion methods should live. `xarray` implements the conversion from / to `pandas`, which `pandas` calls in the `to_xarray` methods. Maybe we can do something similar here?

I told @erogluorhan this before, but I would be open to scheduling a separate, dedicated meeting discussing the finer details if you prefer. Otherwise I'd also be fine with continuing this discussion here or on the `xdggs` repo.

**Comment by erogluorhan** [comment]
Hi @keewis, thanks a lot for opening this discussion, and sorry for my delayed reply (though, I am glad we had a few chances to discuss through Pangeo meetings so far)!
 
UXarray is aimed to support multiple arbitrary, horizontally unstructured grid formats, representing them in a unified, UGRID convention-like model. The essential information Uxarray needs to explicitly have for such a representation is:

* `node_lon`: Longitude of the vertices that build each face/cell
* `node_lat`: Latitude of the vertices that build each face/cell
* `face_node_connectivity`: The connectivity between each face and their vertices, i.e. indices of the corner nodes that make up each face/cell

(We have recently released functionality to create UXarray grids even from point clouds, using Delaunay triangulation and Voronoi tesselations, but it has the connectivity information being created as a result of these processes, hence the same explicit representation)

I think this is the main difference between the UXarray representation and DGGS where there is no need for explicitly defined connectivities because the number and locations of "pixels" for DGGS grids are fixed for each predefined resolution level (and some other related parameters). Each has its own advantages of course, and it'd be great to further our discussions on those ends to better understand each other.

That said, any UXarray representation of a DGGS would require us to generate the connectivity information between faces and nodes. Please see the code snippet in [another discussion](https://github.com/UXARRAY/uxarray/issues/1028#issuecomment-2512344119), which was a preliminary look into recognizing HEALPix grids in UXarray that is nowhere optimized, for getting a sense of it conceptually.

This is also applicable to reading `Xarray.Dataset`s into UXarray. We look into the dataset to parse the grid type (looking into some expected data fields for each grid format) from it. This would not be needed for HEALPix at least because we are planning to add a builtin support for that, but for any other DGGS grid, as long as you can feed UXarray (i.e. via `open_grid()` or `open_dataset()`) with Xarray datasets that have node coords and face-to-node connectivity, UXarray should be able to represent them. Also, we can always discuss to determine if the community would benefit from any other DGGS being natively supported by UXarray, like HEALPix.

I hope this helps and am looking forward to further discussion/coolaboration.

**Reply by keewis** [reply]
coming back to this after drifting off into exploring regridding algorithms, I wonder if we should use `geoarrow` as the interchange format? `xdggs` can generate `geoarrow` from cell ids, so calling `Grid.from_geoarrow` and then converting the dataset to `uxarray` could work.

Otherwise we could also try to define a new `geomesh` format (whatever the name) on top of `arrow` that would consist of the three arrays you talked about (node coords and the face-node connectivity encoded like `geoarrow` polygons minus the ring indices) and use that as the interchange format instead. It would take _a lot_ more work to figure out the details of that, though.

Additionally, if I understand the way you're currently constructing your `Grid` objects from healpix correctly, it is by finding unique vertices and then constructing the face-node connectivity from that, which means you'll have to deal with floating-point accuracy. We should be able to avoid this since any DGGS allows computing the neighbouring cells, and thus it should be possible to derive the connectivity from the integer cell ids (I don't have any idea right now on how to go from local neighbourhood to the face-node connectivity, though). Might be fun to look into, but I'm not sure when I'll have the time.

For converting to `xdggs`, the only thing we'd need is cell ids and the grid parameters. If you don't keep the cell ids around, this would be a lot harder: most DGGS (including healpix) allow combining cells of different sizes into a single grid by encoding the (resolution / zoom) level into the cell id, and thus the usual trick of deriving the cell ids from the cell centroids wouldn't work.

Of these, I think the easiest tasks right now are:
- keep the cell ids in a (lazy) coordinate in `uxarray`
- create `Grid.from_geoarrow` and/or `Grid.from_shapely`
